### ISSUE TITLE
_fe_analyzer_shared v47+ fails to run on web due to bit shift

### ISSUE THREAD

**Issue by ethanblake4** [dart-lang/sdk#50048]
**Edit: ** See https://github.com/dart-lang/sdk/issues/50048#issuecomment-1257269913

`print(-1 << 8)`
prints `-256` in the Dart VM and dart2js. However under DDC it instead prints `4294967040`. This appears to be because DDC is adding an unsigned right shift by 0 (`>>> 0`) in the generated code, not sure why.

(Currently breaking my app in debug mode, thanks for taking a look!)

**Comment by lrhn** [comment]
I believe dart2js does the same, and it's a deliberate and consistent choice when compiling to JavaScript.
If you run the example in dartpad.dev, which is based on dart2js, it also prints `4294967040`.

**Comment by ErikErnst** [comment]
When compiled to JavaScript, Dart uses IEEE-754 double values to represent objects of type `int` as well as `double`. This means that bitwise operations cannot work on a 64 bit value (because the given representation uses only 53 bits to represent the value).

Bitwise operations convert the operands to 32-bit integers (again, just like JavaScript), so you're working on 32-bit values when something like `-1 << 8` is computed.

However, the resulting value is considered to be an _unsigned_ 32-bit two's complement representation, which means that the result returned from `-1 << 8` should be `(-256).toUnsigned(32)` which is 4294967040.

For more detail, check out [this page](https://dart.dev/guides/language/numbers) and p.246 of the [language specification](https://spec.dart.dev/DartLangSpecDraft.pdf).

In any case, the vm uses a 64-bit two's complement representation, and -256 is the expected result, and DDC apparently returns the result which is expected from a web compiler.

But it is a bug if `dart compile js` returns -256.

**Comment by ethanblake4** [comment]
Ah, it does do the same thing in dart2js... perhaps I wasn't hitting that code path due to tree-shaking.

Ok, well in lieu of changing that behavior would it be possible to modify [this code](https://github.com/dart-lang/sdk/blob/cfddeee68e0cf7451a2eee1a9fbbff8a8d9c0b92/pkg/_fe_analyzer_shared/lib/src/scanner/token.dart#L554) in _fe_analyzer_shared, which was added in v0.47 (specifically [this commit](https://github.com/dart-lang/sdk/commit/403b896d580924ddcaec582f2c9584fe41a0e38e)) and worked fine on the Web before that? Or to change [this constant](https://github.com/dart-lang/sdk/blob/cfddeee68e0cf7451a2eee1a9fbbff8a8d9c0b92/pkg/_fe_analyzer_shared/lib/src/scanner/abstract_scanner.dart#L102) to use a positive offset, which I believe would also alleviate the problem.

**Comment by eernstg** [comment]
@scheglov, @chloestefantsova, WDYT?

**Comment by mkustermann** [comment]
If the `offset >= -1`, consider just adding a bias of 1 when encoding:
```diff
diff --git a/pkg/_fe_analyzer_shared/lib/src/scanner/token.dart b/pkg/_fe_analyzer_shared/lib/src/scanner/token.dart
index 0bbb2546695..2bbb26fc1ba 100644
--- a/pkg/_fe_analyzer_shared/lib/src/scanner/token.dart
+++ b/pkg/_fe_analyzer_shared/lib/src/scanner/token.dart
@@ -516,7 +516,7 @@ class SimpleToken implements Token {
    * token.
    */
   @override
-  int get offset => _typeAndOffset >> 8;
+  int get offset => (_typeAndOffset >> 8) - 1;
 
   /**
    * Set the offset from the beginning of the file to the first character in
@@ -525,7 +525,8 @@ class SimpleToken implements Token {
   @override
   void set offset(int value) {
     assert(_tokenTypesByIndex.length < 256);
-    _typeAndOffset = (value << 8) | (_typeAndOffset & 0xff);
+    assert(value >= -1);
+    _typeAndOffset = ((1 + value) << 8) | (_typeAndOffset & 0xff);
   }
 
   /**
@@ -551,8 +552,9 @@ class SimpleToken implements Token {
    * Initialize a newly created token to have the given [type] and [offset].
    */
   SimpleToken(TokenType type, int offset, [this._precedingComment])
-      : _typeAndOffset = ((offset << 8) | type.index) {
+      : _typeAndOffset = (((1 + offset) << 8) | type.index) {
     // Assert the encoding of the [type] is fully reversible.
+    assert(value >= -1);
     assert(type.index < 256 && _tokenTypesByIndex.length < 256);
     assert(identical(offset, this.offset));
     assert(identical(type, this.type));
```

**Comment by blaugold** [comment]
If I'm not mistaken, [Zapp](https://zapp.run/) runs the analyzer in the browser to provided IntelliSense for its code editor.

cc @Salakar

**Comment by ethanblake4** [comment]
My use case is for a code editor I'm building with Flutter as well as [dart_eval](https://pub.dev/packages/dart_eval) (obviously, a bit of a unique case). I wouldn't expect the full analyzer to work on the web, but the main thing I rely on is the parser via the parseString() method which seemed to work fine until the aforementioned commit.

**Comment by scheglov** [comment]
We don't do anything special to prevent running on the web, and theoretically the analyzer can be used with in-memory file system, or a custom implementation of it.

**Comment by jensjoha** [comment]
> We don't do anything special to prevent running on the web, and theoretically the analyzer can be used with in-memory file system, or a custom implementation of it.

There's a big difference between preventing it, and supporting it.

**Comment by scheglov** [comment]
True, I cannot say that we support `package:analyzer` to be run on web, because we don't run any tests for this platform. AFAIK we run tests only using Dart VM.

**Comment by Salakar** [comment]
> My use case is for a code editor I'm building with Flutter as well as [dart_eval](https://pub.dev/packages/dart_eval) (obviously, a bit of a unique case). I wouldn't expect the full analyzer to work on the web, but the main thing I rely on is the parser via the parseString() method which seemed to work fine until the aforementioned commit. edit: actually, it does appear that Zapp runs the full analyzer in the browser, so it does seem to work.

Zapp is currently on the version of `_fe_analyzer_shared` that is shipped with the Dart version in Flutter 3.0.5 since I'm stuck with upgrading to Flutter 3.3.0 and above - there's been a bunch of Analyzer/Compiler internals changes since 3.0.5 that have broken our usage of compiler & analyzer on web completely, I suspect this could be one of them if it's a more recent change but I've not had the time to investigate this one further since there's other issues that I've been stuck on / working on our own patches for.

**Comment by jensjoha** [comment]
I have a fix (@mkustermann's) for this particular issue in review (https://dart-review.googlesource.com/c/sdk/+/261301).
